lib/features/mailbox/data/repository/mailbox_repository_impl.dart

```diff
@@ -1,17 +1,91 @@
+import 'package:core/core.dart';
 import 'package:jmap_dart_client/jmap/account_id.dart';
 import 'package:jmap_dart_client/jmap/core/properties/properties.dart';
+import 'package:jmap_dart_client/jmap/core/state.dart';
 import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox.dart';
 import 'package:tmail_ui_user/features/mailbox/data/datasource/mailbox_datasource.dart';
+import 'package:tmail_ui_user/features/mailbox/data/model/mailbox_cache_response.dart';
+import 'package:tmail_ui_user/features/mailbox/data/model/mailbox_change_response.dart';
+import 'package:tmail_ui_user/features/mailbox/data/model/mailbox_response.dart';
+import 'package:tmail_ui_user/features/mailbox/data/extensions/list_mailbox_cache_extension.dart';
 import 'package:tmail_ui_user/features/mailbox/domain/repository/mailbox_repository.dart';
 
 class MailboxRepositoryImpl extends MailboxRepository {
 
-  final MailboxDataSource mailboxDataSource;
+  final Map<DataSourceType, MailboxDataSource> mapDataSource;
 
-  MailboxRepositoryImpl(this.mailboxDataSource);
+  MailboxRepositoryImpl(this.mapDataSource);
 
   @override
-  Future<List<Mailbox>> getAllMailbox(AccountId accountId, {Properties? properties}) {
-    return mailboxDataSource.getAllMailbox(accountId, properties: properties);
+  Stream<MailboxResponse> getAllMailbox(AccountId accountId, {Properties? properties}) async* {
+    final mailboxCacheResponse = await mapDataSource[DataSourceType.local]!.getAllMailboxCache();
+
+    final mailboxList = mailboxCacheResponse.mailboxCaches != null
+        ? mailboxCacheResponse.mailboxCaches!.toMailboxList()
+        : <Mailbox>[];
+    final oldState = mailboxCacheResponse.oldState?.state;
+
+    yield MailboxResponse(
+      mailboxes: mailboxList,
+      state: oldState != null ? State(oldState) : null);
+
+    if (mailboxList.isNotEmpty && oldState != null) {
+      final changesResponse = await mapDataSource[DataSourceType.network]!.getChanges(accountId, State(oldState));
```

Comment:
How about if changes response `hasMoreChanges = true`?

Reply:
What do you mean ?

Reply:
Please check the specs:
<img width="740" alt="Screen Shot 2021-09-29 at 11 53 03" src="https://user-images.githubusercontent.com/6462404/135205490-6501fb7c-f0b3-4292-a764-c60d9d219d8b.png">

![image](https://user-images.githubusercontent.com/6462404/135205682-dd2adce8-eeb9-4d67-bcd5-bc6ce59af938.png)

if the changes have more than `maxChanges` we need to continue to call to get all the changes

Reply:
I don't understand what you mean. Please give me details.

Reply:
If the total change is 129, but the `/changes` response only have 128, we should do what for this?